### PR TITLE
[WIP] GBTRegressor does not provide uncertainty estimates

### PR DESCRIPTION
**Pull request by betatim**
Started a simple wrapper for regressors like GBTRegressor to provide uncertainty estatimates as well
as central predictions.

Any ideas on how to test that this indeed gives the central and 68% confidence interval (right terminology??) predictions? Best idea so far is to use a very large number of samples on a simple problem and check that the predicted standard deviation is approximately equal to the std used when sampling.

Todo:
* [x] documentation
* [ ] check that GP uses same definition of uncertainty
* [x] smarter tests

### CODE REVIEW

File: skopt/tests/test_gbt.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+def truth(X):
+    return 0.5 * np.sin(1.75*X[:, 0])
+def constant_noise(X):
+    return np.ones_like(X)
+def varying_noise(X):
+    # noise level depends on value of `X`
+    return np.abs(2.5-X)/2.5 + 0.1
+
+def sample_noise(X, std=0.2, noise=constant_noise):
```
Comment (betatim) [inline skopt/tests/test_gbt.py]: should pass in `rng`/use `check_random_state`

File: skopt/tests/test_gbt.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+rng = np.random.RandomState(324)
+
+def truth(X):
+    return 0.5 * np.sin(1.75*X[:, 0])
+
+def constant_noise(X):
+    return np.ones_like(X)
+
+def sample_noise(X, std=0.2, noise=constant_noise):
```
Comment (betatim) [inline skopt/tests/test_gbt.py]: use `random_state` pattern instead of relying on global `rng`

**Comment by glouppe** [comment]
I am not fond of the nomenclature, std != uncertainty != quantile.

**Comment by betatim** [comment]
I'm flexible, but out of ideas at the moment for a good (short) name.

`GBTQuantileRegressor`?

`GBTQuantiles(quantiles=list_of_quantiles)` with `per_quantile_prediction = predict()`. You set all the quantiles you want estimated and we return each single one? In this case `GBTQuantiles([0.18, 0.5, 0.84])`

**Comment by MechCoder** [comment]
I'm planning to review my GradientBoosting know-how, this weekend. Can have a look after that.

**Comment by betatim** [comment]
If you need this merge it, I'm away till Monday night without internet

File: skopt/gbt.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    def fit(self, X, y):
+        """Fit one regressor for each quantile"""
+        rng = check_random_state(self.random_state)
+        self.regressors_ = [GradientBoostingRegressor(loss='quantile',
+                                                      alpha=a,
+                                                      random_state=rng)
+                            for a in self.quantiles]
+        for rgr in self.regressors_:
+            rgr.fit(X, y)
```
Comment (MechCoder) [inline skopt/gbt.py]: return self

**Comment by MechCoder** [comment]
@glouppe @betatim 
How do we compute the standard deviation of the predictions given the quantiles? There is no assumption about the conditional distribution of Y given X right?

**Comment by betatim** [comment]
The last item on the to do list: "check definition of uncertainty" is something I'd like to punt to #23, instead of waiting to merge this till we converge.

**Comment by MechCoder** [comment]
This looks ok to me.

**Comment by betatim** [comment]
:chart_with_upwards_trend:

**Comment by glouppe** [comment]
Would be good to expose the `base_estimator` in this newly introduced class. Default params of `GradientBoostingRegressor` are very likely to give poor results.

**Pull request merged**